dgr/aci-push.go

```diff
@@ -44,21 +52,115 @@ func (aci *Aci) Push() error {
+		queryCheck.Set(key, val)
+	}
+	urlReq.RawQuery = queryCheck.Encode()
+
+	/* Authenticate */
+	req, err := http.NewRequest("HEAD", urlReq.String(), nil)
+	req.SetBasicAuth(Home.Config.Push.Username, Home.Config.Push.Password)
+	resp, err := client.Do(req)
+	if err != nil {
+		logs.WithE(err).Info("Failed to check version")
+		return false, err
+	}
+	ok = false
+	switch {
+	case resp.StatusCode >= 200 && resp.StatusCode <= 300:
+		logs.WithField("statusCode", resp.StatusCode).
+			WithField("status", resp.Status).
+			Info("Aci already uploaded.")
+	case 404 == resp.StatusCode:
+		logs.WithField("statusCode", resp.StatusCode).
+			WithField("status", resp.Status).
+			Info("Ready to upload")
+		ok = true
+	case 401 == resp.StatusCode || 403 == resp.StatusCode:
+		logs.WithField("statusCode", resp.StatusCode).
+			WithField("status", resp.Status).
+			Info("Failed to login to the registry")
+	case resp.StatusCode >= 500:
+		logs.WithField("statusCode", resp.StatusCode).
+			WithField("status", resp.Status).
+			Info("Registry server error")
+	}
+	return
+}
+
+func (aci *Aci) nexusRegistryPush(urlReq url.URL, aciPath string, params map[string]string) (err error) {
+	file, err := os.Open(aciPath)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	body := &bytes.Buffer{}
+	writer := multipart.NewWriter(body)
+	for key, val := range params {
+		_ = writer.WriteField(key, val)
+	}
+	part, err := writer.CreateFormFile("file", filepath.Base(aciPath))
+	if err != nil {
+		return err
+	}
+	_, err = io.Copy(part, file)
+
+	err = writer.Close()
+	if err != nil {
+		return err
+	}
+
+	client := &http.Client{}
+
+	/* Authenticate */
+	req, err := http.NewRequest("POST", urlReq.String(), body)
+	req.SetBasicAuth(Home.Config.Push.Username, Home.Config.Push.Password)
+	req.Header.Set("Content-Type", writer.FormDataContentType())
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	logs.WithField("statusCode", resp.StatusCode).
+		WithField("status", resp.Status).
+		WithField("aciPath", aciPath).
+		Debug("Upload logs")
+	if resp.StatusCode != 201 {
+		return errs.WithE(err, resp.Status)
+	}
+	return
+}
+
 func (aci *Aci) upload(name *common.ACFullname) error {
 	if Home.Config.Push.Type == "maven" && name.DomainName() == "aci.blbl.cr" { // TODO this definitely need to be removed
-		logs.WithF(aci.fields).Info("Uploading aci")
-		if err := common.ExecCmd("curl", "-f", "-i", "-L",
-			"-F", "r=releases",
-			"-F", "hasPom=false",
-			"-F", "e=aci",
-			"-F", "g=com.blablacar.aci.linux.amd64",
-			"-F", "p=aci",
-			"-F", "v="+name.Version(),
-			"-F", "a="+strings.Split(string(name.Name()), "/")[1],
-			"-F", "file=@"+aci.target+pathImageGzAci,
-			"-u", Home.Config.Push.Username+":"+Home.Config.Push.Password,
-			Home.Config.Push.Url+"/service/local/artifact/maven/content"); err != nil {
-			return errs.WithEF(err, aci.fields, "Failed to push aci")
+		urlCheck, _ := url.Parse(Home.Config.Push.Url)
```

Comment:
!!!!

Reply:
Rhoo it can't err. it's not possible ^^
I will fix.

Reply:
I added some error handling.And logs.